cub/thread/thread_operators.cuh

```diff
@@ -37,308 +37,324 @@
 
 #pragma once
 
-#include "../config.cuh"
-#include "../util_type.cuh"
+#include <cub/config.cuh>
+#include <cub/util_type.cuh>
+
+#include <cuda/std/type_traits>
+#include <cuda/std/utility>
 
 CUB_NAMESPACE_BEGIN
 
 
 /**
- * \addtogroup UtilModule
+ * @addtogroup UtilModule
  * @{
  */
 
-/**
- * \brief Default equality functor
- */
+/// @brief Default equality functor
 struct Equality
 {
-    /// Boolean equality operator, returns <tt>(a == b)</tt>
-    template <typename T>
-    __host__ __device__ __forceinline__ bool operator()(const T &a, const T &b) const
-    {
-        return a == b;
-    }
+  /// Boolean equality operator, returns `t == u`
+  template <typename T, typename U>
+  __host__ __device__ __forceinline__ bool operator()(T &&t, U &&u) const
+  {
+    return cuda::std::forward<T>(t) == cuda::std::forward<U>(u);
+  }
```

Comment:
Please tell me that there is nobody out there that specializes comparison operators depending on value categories.

Can we simply bind to a `const T&` and drop the forward?

Applies throughout

Reply:
What's the advantage in this case?

Reply:
Better throughput, we have 2 function calls in `cuda::std::forward`

Also it makes reasoning about the code much more difficult compared to a simple operator that just takes a const reference

Reply:
At some point we are going to alias these operators with libcu++ ones. If we do it later, we'd break behaviour. We can't do it right now, because libcu++ has to backport some specializations to previous dialects. So we'd have to wait till CUB 3.0 to make this change. It's not the case if we mimic the libcu++ behavior now. 

I'm also not sure what kind of throughput change you are referring to. There are no difference in generated [assembly](https://godbolt.org/z/T88xEhdnq) for cases when it doesn't affect behaviour.

Reply:
Yeah sorry, I was referring to compiler throughput. instantiating functions is not entirely free.